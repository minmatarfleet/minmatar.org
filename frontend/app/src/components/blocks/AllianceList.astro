---
import { i18n } from '@helpers/i18n'
const { t } = i18n(Astro.url)

import type { AllianceListData } from '@dtypes/layout_components';

import Flexblock from '@components/compositions/Flexblock.astro';
import FlexInline from '@components/compositions/FlexInline.astro';
import FixedFluid from '@components/compositions/FixedFluid.astro';
import TrashIcon from '@components/icons/buttons/TrashIcon.astro';
import FluidFixed from '@components/compositions/FluidFixed.astro';
import VerticalCenter from '@components/blocks/VerticalCenter.astro';
import Button from '@components/blocks/Button.astro';

interface Props {
    input_name?:        string;
    empty_text?:        string | boolean;
    readonly?:          boolean;
    alliances?:         AllianceListData[];
    [propName: string]: any;
}

const {
    input_name = 'alliances',
    empty_text = t('empty_alliances_text'),
    readonly = true,
    alliances = [],
    ...attributes
} = Astro.props;

delete attributes.class
---

<Flexblock
    class:list={['alliance-list', Astro.props.class]}
    {...attributes}
    gap='var(--space-s)'
    x-data={`{
        alliances: ${JSON.stringify(alliances)},
        alliances_ids: [],
        generate_ids() {
            this.alliances_ids = this.alliances.map(e => e.alliance_id).join(',')
        },
        ${readonly ? '' :
        `add_alliance(alliance) {
            if (!alliance) return

            const exist = this.alliances.find((existing_alliance) => existing_alliance.name === alliance.name)

            if (exist) return
            
            this.alliances.push({
                alt: alliance.name+' ${t('logo').toLowerCase()}',
                name: alliance.name,
                members_count: alliance.members_count,
                alliance_id: alliance.alliance_id,
            })

            this.generate_ids()
        },
        remove_alliance(name) {
            this.alliances = this.alliances.filter(item => item.name !== name)
            this.generate_ids()
        },`}
    }`}
    x-init="generate_ids"
>
    <FlexInline
        justification='space-between'
        gap="var(--space-s)"
    >
        <Flexblock gap="var(--space-3xs)">
            <h2>{t('alliances')}</h2>
            <small class="!mt-[var(--space-3xs)]" x-text={'`${alliances.length} '+t('alliance').toLowerCase()+'${alliances.length != 1 ? `s` : ``}`'}>0 {t('alliances').toLowerCase()}</small>
        </Flexblock>

        {!readonly && 
            <VerticalCenter>
                <Button
                    size='sm'
                    type="button"
                    class="[ w-full ]"
                    x-on:click="add_alliance(await show_alliances_finder())"
                >
                    {t('add_alliance')}
                </Button>
            </VerticalCenter>
        }
    </FlexInline>

    <Flexblock gap='var(--space-3xs)'>
        <template x-for="alliance in alliances">
            <FixedFluid width='64px' class="[ alliance-item ]">
                <picture>
                    <img loading="lazy" x-bind:src="get_alliance_logo(alliance.alliance_id, 256)" width="64" height="64" alt="alliance.alt" />
                </picture>
                <FluidFixed width="46px" breakpoint='60%'>
                    <VerticalCenter>
                        <Flexblock gap='0' class="[ w-full ]">
                            <span class="truncate text-[var(--highlight)]" x-text="alliance.name" />
                            <small class="truncate" x-text={`alliance.members_count+' ${t('members_count').toLowerCase()}'`} />
                        </Flexblock>
                    </VerticalCenter>
                    <VerticalCenter>
                        {!readonly && 
                            <Button
                                class="[ action-button ]"
                                x-on:click="remove_alliance(alliance.name)"
                                size='sm'
                                type="button"
                                narrow={true}
                            >
                                <TrashIcon slot="icon" />
                            </Button>
                        }
                    </VerticalCenter>
                </FluidFixed>
            </FixedFluid>    
        </template>
        
        {empty_text &&
            <p x-bind:class="{ hidden: alliances.length > 0 }">{empty_text}</p>
        }
    </Flexblock>

    <input type="hidden" name={input_name} x-model="alliances_ids">
</Flexblock>

<style lang="scss">
    .alliance-list {
        transition: var(--fast-transition);
        width: 100%;

        img {
            object-fit: cover;
            width: 100%;
            height: 100%;
        }

        .alliance-item {
            @media (hover: hover) {
                .action-button {
                    transition: var(--fast-transition);
                    opacity: 0;
                }

                &:hover,
                &:focus-within {
                    .action-button {
                        opacity: 1;
                    }
                }
            }
        }

        &.disabled {
            pointer-events: none;
            opacity: 0.25;
        }
    }
</style>