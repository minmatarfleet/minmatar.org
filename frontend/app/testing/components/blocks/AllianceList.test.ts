import { experimental_AstroContainer as AstroContainer } from "astro/container";
import { expect, test } from "vitest";
import AllianceList from "@components/blocks/AllianceList.astro";

test("AllianceList defaults", async () => {
  const container = await AstroContainer.create();
  const result = await container.renderToString(AllianceList, {});

  expect(result).toMatchSnapshot();
});
