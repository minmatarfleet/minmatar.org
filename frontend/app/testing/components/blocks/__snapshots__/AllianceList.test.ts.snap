// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AllianceList defaults 1`] = `
"<div class="flex-block alliance-list gap---space-s" x-data="{
        alliances: [],
        alliances_ids: [],
        generate_ids() {
            this.alliances_ids = this.alliances.map(e => e.alliance_id).join(',')
        },
        
    }" x-init="generate_ids" data-astro-cid-zc5ph2vg="true" data-mod style="--flex_block_gap: var(--space-s);">  <div class="flex-inline" data-astro-cid-zc5ph2vg="true" data-astro-cid-ud5fqf5k style="--flex_inline_gap: var(--space-s);--flex_inline_justification: space-between;">  <div class="flex-block gap---space-3xs" data-astro-cid-zc5ph2vg="true" data-mod style="--flex_block_gap: var(--space-3xs);">  <h2 data-astro-cid-zc5ph2vg>Alliances</h2> <small class="!mt-[var(--space-3xs)]" x-text="\`\${alliances.length} alliance\${alliances.length != 1 ? \`s\` : \`\`}\`" data-astro-cid-zc5ph2vg>0 alliances</small>  </div>   </div>   <div class="flex-block gap---space-3xs" data-astro-cid-zc5ph2vg="true" data-mod style="--flex_block_gap: var(--space-3xs);">  <template x-for="alliance in alliances" data-astro-cid-zc5ph2vg> <div class="fixed-fluid [ alliance-item ]" data-astro-cid-zc5ph2vg="true" data-mod data-astro-cid-34go4nvk style="--fixed_fluid_gap: var(--space-s);--fixed_fluid_width: 64px;--fixed_fluid_breakpoint: 50%;">  <picture data-astro-cid-zc5ph2vg> <img loading="lazy" x-bind:src="get_alliance_logo(alliance.alliance_id, 256)" width="64" height="64" alt="alliance.alt" data-astro-cid-zc5ph2vg> </picture> <div class="fluid-fixed" data-astro-cid-zc5ph2vg="true" data-mod data-astro-cid-rwsocfon style="--fixed_fluid_gap: var(--space-s);--fixed_fluid_width: 46px;--fixed_fluid_breakpoint: 60%;">  <div class="!flex !items-center" data-astro-cid-zc5ph2vg="true">  <div class="flex-block [ w-full ] gap-0" data-astro-cid-zc5ph2vg="true" data-mod style="--flex_block_gap: 0;">  <span class="truncate text-[var(--highlight)]" x-text="alliance.name" data-astro-cid-zc5ph2vg></span> <small class="truncate" x-text="alliance.members_count+' members count'" data-astro-cid-zc5ph2vg></small>  </div>   </div> <div class="!flex !items-center" data-astro-cid-zc5ph2vg="true">  </div>  </div>    </div>   </template> <p x-bind:class="{ hidden: alliances.length > 0 }" data-astro-cid-zc5ph2vg>No alliance found.</p> </div>  <input type="hidden" name="alliances" x-model="alliances_ids" data-astro-cid-zc5ph2vg>  </div>  "
`;
